server/src/Persistence.hs

```diff
@@ -69,9 +76,94 @@ apply op val = do
 
 -- * IO
 
+loadFromBackend :: StorageBackend -> PersistenceConfig -> IO (Either String PersistentValue)
+loadFromBackend File pc = loadFile pc
+loadFromBackend Sqlite pc = loadSqliteFile pc
+
+syncToBackend :: StorageBackend -> PersistentValue -> IO ()
+syncToBackend File pv = syncFile pv
+syncToBackend Sqlite pv = syncSqliteFile pv
+
+-- * SQLite loading and syncing
+
+-- | There is currently just one row and it contains only one column of type SBS.ByteString.
+-- This single field holds the whole JSON value for now.
+data JsonRow = JsonRow {jsonByteString :: SBS.ByteString} deriving (Show)
+
+instance FromRow JsonRow where
+  fromRow = JsonRow <$> field
+
+loadSqliteFile :: PersistenceConfig -> IO (Either String PersistentValue)
+loadSqliteFile config = runExceptT $ do
+  liftIO $ forM_ (pcMetrics config) $ \m -> do
+    size <- getFileSize (pcDataFile config)
+    Metrics.setDataSize size m
+
+  value <- readSqliteData (pcDataFile config) (pcLogger config)
+  valueVar <- lift $ newTVarIO value
+  dirtyVar <- lift $ newTVarIO False
+  journal <- for (pcJournalFile config) openJournal
+  let val = PersistentValue
+        { pvConfig  = config
+        , pvValue   = valueVar
+        , pvIsDirty = dirtyVar
+        , pvJournal = journal
+        }
+  recoverJournal val
+  return val
+
+-- | Read and decode the Sqlite data file from disk
+readSqliteData :: FilePath -> Logger -> ExceptT String IO Store.Value
+readSqliteData filePath _logger = ExceptT $ do
+  -- read the data from SQLite
+  conn <- liftIO $ open filePath
+  -- TODO: We might want to have a primary key here after all. Then we could have multiple
+  -- JSON values next to each other.
+  liftIO $ execute_ conn "CREATE TABLE IF NOT EXISTS test (value BLOB)"
+  jsonRows <- liftIO $ (query_ conn "SELECT * from test" :: IO [JsonRow])
+
+  case jsonRows of
+    -- if there is no data yet, we simply return the empty object. We do the same thing for the
+    -- file backend.
+    [] -> do
+      liftIO $ execute conn "INSERT INTO test (value) VALUES (?)" (Only $ Aeson.encode Aeson.emptyObject)
```

Comment:
Also this insert, why do we need it at all? If we don't write the value, we will default to `{}` next time anyway.

If we want to do a dummy insert at startup to ensure that we have write access, I think it is better to do that explicitly during startup. That way we can also deduplicate it between SQLite and the file backend.

Reply:
(Hi; I still am subscribed to this repo apparently. hehe. friendly drive-by comment)

I think this will also maybe cause a race condition.  Should probably rwap the `SELECT * from test` and `INSERT INTO`  into a `BEGIN TRANSACTION` and `COMMIT` no?  otherwise if two people modify the empty database at the same time, you might end up wiping out the changes

Reply:
> otherwise if two people modify the empty database at the same time, you might end up wiping out the changes

Not sure if this is something we need to guard against with the single row single column backend. After all, while icepeak is running it will periodically update the value anyway, without first reading it, and thus overwriting any changes made concurrently.
It's the same as with the other backend: if I modify `icepeak.json` while Icepeak is running, then I cannot expect those changes to be persisted.

If possible, we should make the sqlite connection exclusive so that no concurrent modifications can happen while Icepeak is running.

Reply:
I moved the INSERT to `setupStorageBackend` now.
I agree with Fabian that we assume that Icepeak is the only writer to the Sqlite file - it's the same as for the file backend.